File: appgo/trie.go
package main

type Trie struct {
	Children map[rune]Trie
}

func MakeNew() Trie {
	t := new(Trie)
	t.Children = make(map[rune]Trie)
	return *t
}

func (t *Trie) Insert(word string) {
	curr := *t
	for _, c := range word {
		if !curr.KeyExists(c) {
			curr.Children[c] = MakeNew()
		}
		curr = curr.Children[c]
	}
	curr.Children['$'] = MakeNew()
}

func (t *Trie) KeyExists(c rune) bool {
	curr := *t
	if _, ok := curr.Children[c]; ok {
		return true
	}
	return false
}

func (t *Trie) Search(word string) bool {
	curr := *t
	for _, c := range word {
		if !curr.KeyExists(c) { return false }
		curr = curr.Children[c]
	}
	return curr.KeyExists('$')
}

func (t *Trie) StartsWith(prefix string) bool {
	curr := *t
	for _, c := range prefix {
		if !curr.KeyExists(c) { return false }
		curr = curr.Children[c]
	}
	return true
}

func (t *Trie) AutoCompletions(prefix string) []string {
	return t.AllCompletions()
}

func (t *Trie) AllCompletions() []string {
	completions := make([]string, 0)
	curr := *t
	for k := range curr.Children {
		if k == '$' { continue }
		child := curr.Children[k]
		childCompletions := child.AllCompletions()
		if len(childCompletions) > 0 {
			for _, childCompl := range childCompletions {
				completions = append(completions, string(k) + childCompl)
			}
		} else {
			completions = append(completions, string(k))
		}
	}
	return completions
}

File: app/trie.py

class Trie:

    """ standard methods: insert, search, startsWith """

    def __init__(self):
        self.d = {}

    def insert(self, word):
        curr = self.d
        for c in word:
            if c not in curr:
                curr[c] = {}
            curr = curr[c]
        curr['$'] = {}  # end of word marker

    def search(self, word):
        curr = self.d
        for c in word:
            if c not in curr: return False
            curr = curr[c]
        return '$' in curr

    def startswith(self, word):
        curr = self.d
        for c in word:
            if c not in curr: return False
            curr = curr[c]
        return True

    def autocompletions(self, prefix):
        # an additional method: autocompletion
        completions = []
        curr = self.d
        for c in prefix:
            if c not in curr: return completions
            curr = curr[c]
        # dfs rest of tree
        completions = self.__all_completions(curr)
        return [prefix + cpl for cpl in completions]

    def __all_completions(self, curr):
        # (also an additional method)
        # should make iterative instead of recursive
        completions = []
        for k in curr:
            if k == '$': continue
            child_completions = self.__all_completions(curr[k])
            if child_completions:
                for child_c in child_completions:
                    completions.append(k + child_c)
            else:
                completions.append(k)
        return completions

    def serialize(self):
        enc = []

        def pre_enc(c, curr):
            enc.append(c)
            if c != '$':
                for ch in curr:
                    pre_enc(ch, curr[ch])

        pre_enc('', self.d)
        return ''.join(enc)

    def deserialize(self, data):
        enc = list(data)
        # big problem: can't serialize /deserialized preorder if don't know
        # number of children for a given node.
        def pre_dec(enc):
            c = enc.pop(0)
            if c == '$':
                return {'$': {}}
            curr = { c: {}}
            left = pre_dec(enc) # this isn't correct, there are multiple children
            right = pre_dec(enc)
            curr[c] = {**left, **right}
            return curr

        return pre_dec(enc)


